src/main/kotlin/app/revanced/patcher/Patcher.kt

```diff
@@ -191,201 +67,138 @@ class Patcher(private val options: PatcherOptions) {
-                    executedPatches[patchName] = ExecutedPatch(patchInstance, it.isSuccess())
-                }
-            } catch (e: Exception) {
-                PatchResultError(e).also {
-                    executedPatches[patchName] = ExecutedPatch(patchInstance, false)
-                }
+            var success = false
+            try {
+                patchInstance.execute(patchContext)
+
+                success = true
+            } catch (patchException: PatchException) {
+                throw patchException
+            } catch (throwable: Throwable) {
+                throw PatchException("Unhandled patch exception: ${throwable.message}", throwable)
+            } finally {
+                executedPatches[patchName] = ExecutedPatch(patchInstance, success)
             }
         }
+        if (mergeIntegrations) context.integrations.merge(logger)
 
-        return sequence {
-            if (mergeIntegrations) context.integrations.merge(logger)
-
-            // prevent from decoding the manifest twice if it is not needed
-            if (resourceDecodingMode == ResourceDecodingMode.FULL) decodeResources(ResourceDecodingMode.FULL)
-
-            logger.trace("Executing all patches")
-
-            val executedPatches = LinkedHashMap<String, ExecutedPatch>() // first is name
+        logger.trace("Executing all patches")
 
+        HashMap<String, ExecutedPatch>().apply {
             try {
                 context.patches.forEach { patch ->
-                    val patchResult = executePatch(patch, executedPatches)
-
-                    val result = if (patchResult.isSuccess()) {
-                        Result.success(patchResult.success()!!)
-                    } else {
-                        Result.failure(patchResult.error()!!)
+                    var exception: PatchException? = null
+                    try {
+                        executePatch(patch, this)
+                    } catch (patchException: PatchException) {
+                        exception = patchException
                     }
 
-                    yield(patch.patchName to result)
-                    if (stopOnError && patchResult.isError()) return@sequence
+                    yield(patch.patchName to exception)
+                    if (stopOnError && exception != null) return@sequence
                 }
             } finally {
-                executedPatches.values.reversed().forEach { (patch, _) ->
+                values.reversed().forEach { (patch, _) ->
                     patch.close()
                 }
             }
         }
     }
 
     /**
-     * The type of decoding the resources.
+     * Write patched [Apk]s to a folder.
+     *
+     * @param folder The output folder.
+     * @return The [PatcherResult] of the [Patcher].
      */
-    private enum class ResourceDecodingMode {
-        /**
-         * Decode all resources.
-         */
-        FULL,
+    fun write(folder: File): PatcherResult {
```

Comment:
This function should return the files and not consume a folder.

Reply:
I changed it to be similar to the implementation in split branch.